File: src/bindings/zig/corsa_utils.zig
const std = @import("std");

pub const c = @cImport({
    @cInclude("corsa_utils.h");
});

pub fn toRef(text: []const u8) c.CorsaStrRef {
    return .{
        .ptr = if (text.len == 0) null else text.ptr,
        .len = text.len,
    };
}

fn withRefs(
    allocator: std.mem.Allocator,
    values: []const []const u8,
    body: fn ([]c.CorsaStrRef) bool,
) !bool {
    const refs = try allocator.alloc(c.CorsaStrRef, values.len);
    defer allocator.free(refs);
    for (values, 0..) |value, index| {
        refs[index] = toRef(value);
    }
    return body(refs);
}

fn withDualRefs(
    allocator: std.mem.Allocator,
    left: []const []const u8,
    right: []const []const u8,
    body: fn ([]c.CorsaStrRef, []c.CorsaStrRef) bool,
) !bool {
    const left_refs = try allocator.alloc(c.CorsaStrRef, left.len);
    defer allocator.free(left_refs);
    for (left, 0..) |value, index| left_refs[index] = toRef(value);
    const right_refs = try allocator.alloc(c.CorsaStrRef, right.len);
    defer allocator.free(right_refs);
    for (right, 0..) |value, index| right_refs[index] = toRef(value);
    return body(left_refs, right_refs);
}

pub fn takeString(allocator: std.mem.Allocator, value: c.CorsaString) ![]u8 {
    defer c.corsa_utils_string_free(value);
    if (value.ptr == null or value.len == 0) {
        return allocator.alloc(u8, 0);
    }
    const slice = @as([*]const u8, @ptrCast(value.ptr))[0..value.len];
    return allocator.dupe(u8, slice);
}

pub fn takeStringList(allocator: std.mem.Allocator, value: c.CorsaStringList) ![][]u8 {
    defer c.corsa_utils_string_list_free(value);
    const out = try allocator.alloc([]u8, value.len);
    for (0..value.len) |index| {
        const item = value.ptr[index];
        const slice = if (item.ptr == null or item.len == 0)
            &[_]u8{}
        else
            @as([*]const u8, @ptrCast(item.ptr))[0..item.len];
        out[index] = try allocator.dupe(u8, slice);
    }
    return out;
}

pub fn classifyTypeText(allocator: std.mem.Allocator, text: []const u8) ![]u8 {
    return takeString(allocator, c.corsa_utils_classify_type_text(toRef(text)));
}

pub fn splitTopLevelTypeText(
    allocator: std.mem.Allocator,
    text: []const u8,
    delimiter: u8,
) ![][]u8 {
    return takeStringList(
        allocator,
        c.corsa_utils_split_top_level_type_text(toRef(text), @as(c_uint, delimiter)),
    );
}

pub fn splitTypeText(allocator: std.mem.Allocator, text: []const u8) ![][]u8 {
    return takeStringList(allocator, c.corsa_utils_split_type_text(toRef(text)));
}

pub fn isStringLikeTypeTexts(allocator: std.mem.Allocator, type_texts: []const []const u8) !bool {
    return withRefs(allocator, type_texts, struct {
        fn call(refs: []c.CorsaStrRef) bool {
            return c.corsa_utils_is_string_like_type_texts(
                if (refs.len == 0) null else refs.ptr,
                refs.len,
            );
        }
    }.call);
}

pub fn isNumberLikeTypeTexts(allocator: std.mem.Allocator, type_texts: []const []const u8) !bool {
    return withRefs(allocator, type_texts, struct {
        fn call(refs: []c.CorsaStrRef) bool {
            return c.corsa_utils_is_number_like_type_texts(
                if (refs.len == 0) null else refs.ptr,
                refs.len,
            );
        }
    }.call);
}

pub fn isBigintLikeTypeTexts(allocator: std.mem.Allocator, type_texts: []const []const u8) !bool {
    return withRefs(allocator, type_texts, struct {
        fn call(refs: []c.CorsaStrRef) bool {
            return c.corsa_utils_is_bigint_like_type_texts(
                if (refs.len == 0) null else refs.ptr,
                refs.len,
            );
        }
    }.call);
}

pub fn isAnyLikeTypeTexts(allocator: std.mem.Allocator, type_texts: []const []const u8) !bool {
    return withRefs(allocator, type_texts, struct {
        fn call(refs: []c.CorsaStrRef) bool {
            return c.corsa_utils_is_any_like_type_texts(if (refs.len == 0) null else refs.ptr, refs.len);
        }
    }.call);
}

pub fn isUnknownLikeTypeTexts(allocator: std.mem.Allocator, type_texts: []const []const u8) !bool {
    return withRefs(allocator, type_texts, struct {
        fn call(refs: []c.CorsaStrRef) bool {
            return c.corsa_utils_is_unknown_like_type_texts(
                if (refs.len == 0) null else refs.ptr,
                refs.len,
            );
        }
    }.call);
}

pub fn isArrayLikeTypeTexts(allocator: std.mem.Allocator, type_texts: []const []const u8) !bool {
    return withRefs(allocator, type_texts, struct {
        fn call(refs: []c.CorsaStrRef) bool {
            return c.corsa_utils_is_array_like_type_texts(
                if (refs.len == 0) null else refs.ptr,
                refs.len,
            );
        }
    }.call);
}

pub fn isPromiseLikeTypeTexts(
    allocator: std.mem.Allocator,
    type_texts: []const []const u8,
    property_names: []const []const u8,
) !bool {
    return withDualRefs(allocator, type_texts, property_names, struct {
        fn call(type_refs: []c.CorsaStrRef, property_refs: []c.CorsaStrRef) bool {
            return c.corsa_utils_is_promise_like_type_texts(
                if (type_refs.len == 0) null else type_refs.ptr,
                type_refs.len,
                if (property_refs.len == 0) null else property_refs.ptr,
                property_refs.len,
            );
        }
    }.call);
}

pub fn isErrorLikeTypeTexts(
    allocator: std.mem.Allocator,
    type_texts: []const []const u8,
    property_names: []const []const u8,
) !bool {
    return withDualRefs(allocator, type_texts, property_names, struct {
        fn call(type_refs: []c.CorsaStrRef, property_refs: []c.CorsaStrRef) bool {
            return c.corsa_utils_is_error_like_type_texts(
                if (type_refs.len == 0) null else type_refs.ptr,
                type_refs.len,
                if (property_refs.len == 0) null else property_refs.ptr,
                property_refs.len,
            );
        }
    }.call);
}

pub fn hasUnsafeAnyFlow(
    allocator: std.mem.Allocator,
    source_texts: []const []const u8,
    target_texts: []const []const u8,
) !bool {
    return withDualRefs(allocator, source_texts, target_texts, struct {
        fn call(source_refs: []c.CorsaStrRef, target_refs: []c.CorsaStrRef) bool {
            return c.corsa_utils_has_unsafe_any_flow(
                if (source_refs.len == 0) null else source_refs.ptr,
                source_refs.len,
                if (target_refs.len == 0) null else target_refs.ptr,
                target_refs.len,
            );
        }
    }.call);
}

pub fn isUnsafeAssignment(
    allocator: std.mem.Allocator,
    source_texts: []const []const u8,
    target_texts: []const []const u8,
) !bool {
    return withDualRefs(allocator, source_texts, target_texts, struct {
        fn call(source_refs: []c.CorsaStrRef, target_refs: []c.CorsaStrRef) bool {
            return c.corsa_utils_is_unsafe_assignment(
                if (source_refs.len == 0) null else source_refs.ptr,
                source_refs.len,
                if (target_refs.len == 0) null else target_refs.ptr,
                target_refs.len,
            );
        }
    }.call);
}

pub fn isUnsafeReturn(
    allocator: std.mem.Allocator,
    source_texts: []const []const u8,
    target_texts: []const []const u8,
) !bool {
    return withDualRefs(allocator, source_texts, target_texts, struct {
        fn call(source_refs: []c.CorsaStrRef, target_refs: []c.CorsaStrRef) bool {
            return c.corsa_utils_is_unsafe_return(
                if (source_refs.len == 0) null else source_refs.ptr,
                source_refs.len,
                if (target_refs.len == 0) null else target_refs.ptr,
                target_refs.len,
            );
        }
    }.call);
}

File: src/bindings/swift/CorsaUtils/Sources/CorsaUtils/CorsaUtils.swift
import Foundation

public struct CorsaStrRef {
    public let ptr: UnsafePointer<UInt8>?
    public let len: Int
}

public struct CorsaString {
    public let ptr: UnsafeMutablePointer<CChar>?
    public let len: Int
}

public struct CorsaBytes {
    public let ptr: UnsafeMutablePointer<UInt8>?
    public let len: Int
    public let present: Bool
}

public struct CorsaStringList {
    public let ptr: UnsafeMutablePointer<CorsaString>?
    public let len: Int
}

@_silgen_name("corsa_utils_classify_type_text")
private func classifyTypeTextNative(_ text: CorsaStrRef) -> CorsaString

@_silgen_name("corsa_utils_split_top_level_type_text")
private func splitTopLevelTypeTextNative(_ text: CorsaStrRef, _ delimiter: UInt32) -> CorsaStringList

@_silgen_name("corsa_utils_split_type_text")
private func splitTypeTextNative(_ text: CorsaStrRef) -> CorsaStringList

@_silgen_name("corsa_utils_is_string_like_type_texts")
private func isStringLikeTypeTextsNative(_ typeTexts: UnsafePointer<CorsaStrRef>?, _ typeTextsLen: Int) -> Bool

@_silgen_name("corsa_utils_is_number_like_type_texts")
private func isNumberLikeTypeTextsNative(_ typeTexts: UnsafePointer<CorsaStrRef>?, _ typeTextsLen: Int) -> Bool

@_silgen_name("corsa_utils_is_bigint_like_type_texts")
private func isBigintLikeTypeTextsNative(_ typeTexts: UnsafePointer<CorsaStrRef>?, _ typeTextsLen: Int) -> Bool

@_silgen_name("corsa_utils_is_any_like_type_texts")
private func isAnyLikeTypeTextsNative(_ typeTexts: UnsafePointer<CorsaStrRef>?, _ typeTextsLen: Int) -> Bool

@_silgen_name("corsa_utils_is_unknown_like_type_texts")
private func isUnknownLikeTypeTextsNative(_ typeTexts: UnsafePointer<CorsaStrRef>?, _ typeTextsLen: Int) -> Bool

@_silgen_name("corsa_utils_is_array_like_type_texts")
private func isArrayLikeTypeTextsNative(_ typeTexts: UnsafePointer<CorsaStrRef>?, _ typeTextsLen: Int) -> Bool

@_silgen_name("corsa_utils_is_promise_like_type_texts")
private func isPromiseLikeTypeTextsNative(
    _ typeTexts: UnsafePointer<CorsaStrRef>?,
    _ typeTextsLen: Int,
    _ propertyNames: UnsafePointer<CorsaStrRef>?,
    _ propertyNamesLen: Int
) -> Bool

@_silgen_name("corsa_utils_is_error_like_type_texts")
private func isErrorLikeTypeTextsNative(
    _ typeTexts: UnsafePointer<CorsaStrRef>?,
    _ typeTextsLen: Int,
    _ propertyNames: UnsafePointer<CorsaStrRef>?,
    _ propertyNamesLen: Int
) -> Bool

@_silgen_name("corsa_utils_has_unsafe_any_flow")
private func hasUnsafeAnyFlowNative(
    _ sourceTexts: UnsafePointer<CorsaStrRef>?,
    _ sourceTextsLen: Int,
    _ targetTexts: UnsafePointer<CorsaStrRef>?,
    _ targetTextsLen: Int
) -> Bool

@_silgen_name("corsa_utils_is_unsafe_assignment")
private func isUnsafeAssignmentNative(
    _ sourceTexts: UnsafePointer<CorsaStrRef>?,
    _ sourceTextsLen: Int,
    _ targetTexts: UnsafePointer<CorsaStrRef>?,
    _ targetTextsLen: Int
) -> Bool

@_silgen_name("corsa_utils_is_unsafe_return")
private func isUnsafeReturnNative(
    _ sourceTexts: UnsafePointer<CorsaStrRef>?,
    _ sourceTextsLen: Int,
    _ targetTexts: UnsafePointer<CorsaStrRef>?,
    _ targetTextsLen: Int
) -> Bool

@_silgen_name("corsa_utils_string_free")
func freeStringNative(_ value: CorsaString)

@_silgen_name("corsa_bytes_free")
func freeBytesNative(_ value: CorsaBytes)

@_silgen_name("corsa_utils_string_list_free")
func freeStringListNative(_ value: CorsaStringList)

public enum CorsaUtils {
    public static func classifyTypeText(_ text: String) -> String {
        withStrRef(text) { takeString(classifyTypeTextNative($0)) }
    }

    public static func splitTopLevelTypeText(_ text: String, delimiter: UnicodeScalar) -> [String] {
        withStrRef(text) { takeStringList(splitTopLevelTypeTextNative($0, delimiter.value)) }
    }

    public static func splitTypeText(_ text: String) -> [String] {
        withStrRef(text) { takeStringList(splitTypeTextNative($0)) }
    }

    public static func isStringLikeTypeTexts(_ typeTexts: [String]) -> Bool { callSingle(typeTexts, isStringLikeTypeTextsNative) }
    public static func isNumberLikeTypeTexts(_ typeTexts: [String]) -> Bool { callSingle(typeTexts, isNumberLikeTypeTextsNative) }
    public static func isBigintLikeTypeTexts(_ typeTexts: [String]) -> Bool { callSingle(typeTexts, isBigintLikeTypeTextsNative) }
    public static func isAnyLikeTypeTexts(_ typeTexts: [String]) -> Bool { callSingle(typeTexts, isAnyLikeTypeTextsNative) }
    public static func isUnknownLikeTypeTexts(_ typeTexts: [String]) -> Bool { callSingle(typeTexts, isUnknownLikeTypeTextsNative) }
    public static func isArrayLikeTypeTexts(_ typeTexts: [String]) -> Bool { callSingle(typeTexts, isArrayLikeTypeTextsNative) }

    public static func isPromiseLikeTypeTexts(_ typeTexts: [String], propertyNames: [String] = []) -> Bool {
        callDual(typeTexts, propertyNames, isPromiseLikeTypeTextsNative)
    }

    public static func isErrorLikeTypeTexts(_ typeTexts: [String], propertyNames: [String] = []) -> Bool {
        callDual(typeTexts, propertyNames, isErrorLikeTypeTextsNative)
    }

    public static func hasUnsafeAnyFlow(_ sourceTexts: [String], targetTexts: [String]) -> Bool {
        callFlow(sourceTexts, targetTexts, hasUnsafeAnyFlowNative)
    }

    public static func isUnsafeAssignment(_ sourceTexts: [String], targetTexts: [String]) -> Bool {
        callFlow(sourceTexts, targetTexts, isUnsafeAssignmentNative)
    }

    public static func isUnsafeReturn(_ sourceTexts: [String], targetTexts: [String]) -> Bool {
        callFlow(sourceTexts, targetTexts, isUnsafeReturnNative)
    }
}

func callSingle(_ typeTexts: [String], _ body: (UnsafePointer<CorsaStrRef>?, Int) -> Bool) -> Bool {
    withRefs(typeTexts) { body($0.baseAddress, $0.count) }
}

func callDual(
    _ typeTexts: [String],
    _ propertyNames: [String],
    _ body: (UnsafePointer<CorsaStrRef>?, Int, UnsafePointer<CorsaStrRef>?, Int) -> Bool
) -> Bool {
    withRefs(typeTexts) { typeRefs in
        withRefs(propertyNames) { propertyRefs in
            body(typeRefs.baseAddress, typeRefs.count, propertyRefs.baseAddress, propertyRefs.count)
        }
    }
}

func callFlow(
    _ sourceTexts: [String],
    _ targetTexts: [String],
    _ body: (UnsafePointer<CorsaStrRef>?, Int, UnsafePointer<CorsaStrRef>?, Int) -> Bool
) -> Bool {
    withRefs(sourceTexts) { sourceRefs in
        withRefs(targetTexts) { targetRefs in
            body(sourceRefs.baseAddress, sourceRefs.count, targetRefs.baseAddress, targetRefs.count)
        }
    }
}

func withStrRef<T>(_ value: String, _ body: (CorsaStrRef) -> T) -> T {
    let refs = BorrowedRefs([value])
    return refs.refs.withUnsafeBufferPointer { body($0.first ?? CorsaStrRef(ptr: nil, len: 0)) }
}

func withRefs<T>(_ values: [String], _ body: (UnsafeBufferPointer<CorsaStrRef>) -> T) -> T {
    let refs = BorrowedRefs(values)
    return refs.refs.withUnsafeBufferPointer(body)
}

func takeString(_ value: CorsaString) -> String {
    defer { freeStringNative(value) }
    guard let ptr = value.ptr, value.len > 0 else {
        return ""
    }
    return String(decoding: UnsafeBufferPointer(start: UnsafeRawPointer(ptr).assumingMemoryBound(to: UInt8.self), count: value.len), as: UTF8.self)
}

func takeStringList(_ value: CorsaStringList) -> [String] {
    defer { freeStringListNative(value) }
    guard let ptr = value.ptr, value.len > 0 else {
        return []
    }
    return UnsafeBufferPointer(start: ptr, count: value.len).map { item in
        guard let ptr = item.ptr, item.len > 0 else {
            return ""
        }
        return String(decoding: UnsafeBufferPointer(start: UnsafeRawPointer(ptr).assumingMemoryBound(to: UInt8.self), count: item.len), as: UTF8.self)
    }
}

func takeBytes(_ value: CorsaBytes) -> Data? {
    defer { freeBytesNative(value) }
    guard value.present else {
        return nil
    }
    guard let ptr = value.ptr, value.len > 0 else {
        return Data()
    }
    return Data(bytes: ptr, count: value.len)
}

final class BorrowedRefs {
    private let storage: [UnsafeMutableBufferPointer<UInt8>]
    let refs: [CorsaStrRef]

    init(_ values: [String]) {
        var storage = [UnsafeMutableBufferPointer<UInt8>]()
        var refs = [CorsaStrRef]()
        storage.reserveCapacity(values.count)
        refs.reserveCapacity(values.count)
        for value in values {
            let bytes = Array(value.utf8)
            guard !bytes.isEmpty else {
                refs.append(CorsaStrRef(ptr: nil, len: 0))
                continue
            }
            let buffer = UnsafeMutableBufferPointer<UInt8>.allocate(capacity: bytes.count)
            _ = buffer.initialize(from: bytes)
            storage.append(buffer)
            refs.append(CorsaStrRef(ptr: UnsafePointer(buffer.baseAddress!), len: bytes.count))
        }
        self.storage = storage
        self.refs = refs
    }

    deinit {
        for buffer in storage {
            buffer.deallocate()
        }
    }
}
